airflow/models/dagrun.py

```diff
@@ -533,6 +533,25 @@ def get_previous_scheduled_dagrun(self, session: Session = NEW_SESSION) -> DagRu
             .first()
         )
 
+    def _tis_for_dagrun_state(self, *, dag, tis):
+        """
+        Return the collection of tasks that should be considered for evaluation of terminal dag run state.
+
+        Teardown tasks by default are not considered for the purpose of dag run state.  But
+        users may enable such consideration with on_failure_fail_dagrun.
+        """
+        leaf_task_ids = {t.task_id for t in dag.leaves}
+        leaf_tis = {ti for ti in tis if ti.task_id in leaf_task_ids if ti.state != TaskInstanceState.REMOVED}
+        if dag.teardowns:
```

Comment:
Can use an early return instead

Reply:
I explored this but doing so adds an extra return statement which is not needed if we do it this way.